core/src/main/java/org/mule/api/metadata/DefaultMetadataCache.java

```diff
@@ -0,0 +1,36 @@
+/*
+ * Copyright (c) MuleSoft, Inc.  All rights reserved.  http://www.mulesoft.com
+ * The software in this package is published under the terms of the CPAL v1.0
+ * license, a copy of which has been included with this distribution in the
+ * LICENSE.txt file.
+ */
+package org.mule.api.metadata;
+
+import java.io.Serializable;
+import java.util.HashMap;
+import java.util.Map;
+import java.util.Optional;
+
+final class DefaultMetadataCache implements MetadataCache
+{
+
+    private final Map<Serializable, Serializable> cache = new HashMap<>();
```

Comment:
maybe we can parameterize only the get method to avoid casting, but keep Serializable internally.